### PR TITLE
Don't persist type information to translog

### PR DESCRIPTION
**Pull request by romseygeek**
We no longer need to store type information in the translog, given that an index
can only have a single type.

Relates to #41059

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/translog/Translog.java [hunk 1]
```diff
@@ -1219,7 +1217,9 @@ private void write(final StreamOutput out) throws IOException {
             final int format = SERIALIZATION_FORMAT;
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/index/translog/Translog.java]: I think we need to use the old format for an old wire version. This should be:
```java
final int format = out.getVersion().onOrAfter(Version.V_8_0_0) ? SERIALIZATION_FORMAT : FORMAT_NO_VERSION_TYPE;
```

File: server/src/main/java/org/elasticsearch/index/translog/Translog.java [hunk 2]
```diff
@@ -1381,7 +1377,6 @@ public Source getSource() {
         private void write(final StreamOutput out) throws IOException {
             final int format = SERIALIZATION_FORMAT;
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/index/translog/Translog.java]: Same here. I think we need to use the old format for an old wire version. This should be:

```java
final int format = out.getVersion().onOrAfter(Version.V_8_0_0) ? SERIALIZATION_FORMAT : FORMAT_NO_VERSION_TYPE;
```

**Review by dnhatn — COMMENTED** [review]
Looks great. Thanks, @romseygeek.

I left two comments relating to BWC in translog. I will need some time to look carefully at other files.

File: test/framework/src/main/java/org/elasticsearch/index/engine/TranslogHandler.java [hunk 3]
```diff
@@ -83,7 +83,7 @@ private void applyOperation(Engine engine, Engine.Operation operation) throws IO
                 Engine.Index engineIndex = (Engine.Index) operation;
                 Mapping update = engineIndex.parsedDoc().dynamicMappingsUpdate();
                 if (engineIndex.parsedDoc().dynamicMappingsUpdate() != null) {
-                    recoveredTypes.compute(engineIndex.type(), (k, mapping) -> mapping == null ? update : mapping.merge(update));
+                    recoveredTypes.compute("", (k, mapping) -> mapping == null ? update : mapping.merge(update));
```
Comment (dnhatn) [inline test/framework/src/main/java/org/elasticsearch/index/engine/TranslogHandler.java]: I think we should remove "recoveredTypes" and the related tests.

**Review by dnhatn — APPROVED** [review]
LGTM. Thanks, @romseygeek for your huge effort. I left two smaller comments.

**Comment by romseygeek** [comment]
Thanks @dnhatn, I pushed a change removing `recoveredTypes`, and also `TranslogHandler.mappingUpdate` which only seemed to be set by tests.

**Comment by romseygeek** [comment]
@elasticmachine update branch

**Pull request merged**